src/services/identity/UsersService.ts

```diff
@@ -208,201 +212,169 @@ class UsersService {
         if (!otpEntry) {
           return errAsync(new BadRequestError("OTP not found"))
         }
 
-        return okAsync(otpEntry)
-      })
-      .andThen((otpDbEntry) => {
-        if (otpDbEntry.attempts >= MAX_NUM_OTP_ATTEMPTS) {
+        // after this point, otpEntry is guaranteed to be truthy is all promise chain steps (just like transaction)
+
+        // verify otpEntry validity
+
+        if (otpEntry.expiresAt < new Date()) {
+          return this.otpDestroyEntry(otpEntry, transaction!).andThen(() =>
+            errAsync(new BadRequestError("OTP has expired"))
+          )
+        }
+
+        if (otpEntry.attempts >= MAX_NUM_OTP_ATTEMPTS) {
+          // should this delete the otpEntry as well?
           return errAsync(new BadRequestError("Max number of attempts reached"))
         }
 
-        return okAsync(otpDbEntry)
-      })
-      .andThen((otpDbEntry) => {
-        if (!otpDbEntry.hashedOtp) {
-          return ResultAsync.fromPromise(
-            otpDbEntry.destroy({ transaction }),
-            (error) => {
-              logger.error(
-                `Error destroying OTP entry: ${JSON.stringify(error)}`
-              )
-
-              return new DatabaseError("Error destroying OTP entry in database")
-            }
-          ).andThen(() => errAsync(new BadRequestError("Hashed OTP not found")))
+        if (!otpEntry.hashedOtp) {
+          return this.otpDestroyEntry(otpEntry, transaction!).andThen(() =>
+            errAsync(new BadRequestError("Hashed OTP not found"))
+          )
         }
 
-        return okAsync(otpDbEntry)
-      })
-      .andThen((otpDbEntry) =>
-        // increment attempts
-        ResultAsync.fromPromise(
+        return ResultAsync.fromPromise(
           this.otpRepository.increment("attempts", {
-            where: { id: otpDbEntry.id },
+            where: { id: otpEntry.id },
             transaction,
           }),
-          (error) => {
-            logger.error(
-              `Error incrementing OTP attempts: ${JSON.stringify(error)}`
+          (error) =>
+            this.otpGetAndLogError(
+              DatabaseError,
+              error,
+              "Error incrementing OTP attempts"
             )
-
-            return new DatabaseError("Error incrementing OTP attempts")
-          }
-        ).map(() => otpDbEntry)
-      )
-      .andThen((otpDbEntry) =>
-        ResultAsync.combine([
-          okAsync(otpDbEntry),
-          ResultAsync.fromPromise(
-            this.otpService.verifyOtp(otp, otpDbEntry.hashedOtp),
-            (error) => {
-              logger.error(`Error verifying OTP: ${JSON.stringify(error)}`)
-
-              return new BadRequestError("Error verifying OTP")
-            }
-          ),
-        ])
+        )
+      })
+      .andThen(() =>
+        ResultAsync.fromPromise(
+          this.otpService.verifyOtp(otp, otpEntry!.hashedOtp),
+          (error) =>
+            this.otpGetAndLogError(
+              BadRequestError,
```

Comment:
⚠️ This should not be a `BadRequestError`. Failure to perform the verification is not the same as failing the verification (which is done by checking the boolean output at the next step).

Failure to perform the verification should be an unexpected internal server error, maybe we could just use `BaseIsomerError` here? Or make an new Error type 🤔